File: src/main/java/info/rsdev/playlists/services/MusicChartsService.java
/*
 * Copyright 2018 Red Star Development.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package info.rsdev.playlists.services;

import java.time.LocalDate;
import java.time.temporal.ChronoField;
import java.util.ArrayList;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

import info.rsdev.playlists.dao.ChartsItemDao;
import info.rsdev.playlists.dao.Initializable;
import info.rsdev.playlists.domain.Song;
import info.rsdev.playlists.exception.FailedHostException;

/**
 * This is the default implementation of a {@link MusicTitleService}
 */
@Component
public class MusicChartsService implements MusicTitleService {

    private static final Logger LOGGER = LoggerFactory.getLogger(MusicChartsService.class);

    private final ScrapeService scrapeService;

    private final ChartsItemDao chartsItemDao;

    public MusicChartsService(ScrapeService scrapeService, ChartsItemDao chartsItemDao) {
        this.scrapeService = scrapeService;
        this.chartsItemDao = chartsItemDao;
    }

    @Override
    public void init() {
        if (chartsItemDao instanceof Initializable init && init.setupStoreWhenNeeded()) {
            LOGGER.info("New datastore created succesfully");
        }

        for (MusicChart chart : MusicChart.values()) {
            var maxYear = chartsItemDao.getHighestYearStored(chart);
            if (maxYear >= 0) {
                var maxWeek = chartsItemDao.getHighestWeekStored(chart, maxYear);
                LOGGER.warn("Datastore contains data for {} from {}, week {}", chart, maxYear, maxWeek);
            }
        }
        loadData();
    }

    @Override
    public List<Song> getChartedSongsForYear(short year) {
        return new ArrayList<>(chartsItemDao.getReleases(year));
    }

    /**
     * Complete the database with data from the missing editions of the wanted
     * charts by retrieving them from the Internet via the {@link ScrapeService}
     */
    private void loadData() {
        short currentYear = (short) LocalDate.now().getYear();
        for (MusicChart chart : scrapeService.getSupportedCharts()) {
            try {
                for (short year = getFirstYearToScrape(chart); year <= currentYear; year++) {
                    for (byte weekNumber = firstWeek(chart, year); weekNumber <= lastWeek(currentYear, year); weekNumber++) {
                        long start = System.currentTimeMillis();
                        var chartsItems = scrapeService.scrape(newInternetDocumentFetcher(chart, year, weekNumber));
                        if (!chartsItems.isEmpty()) {
                            long scrapeMillies = System.currentTimeMillis() - start;
                            chartsItemDao.insert(chartsItems);
                            long dbMillies = System.currentTimeMillis() - start - scrapeMillies;
                            LOGGER.info("persisted week {} of {} {} ({} songs [ scrape={}ms., db={}ms.])", weekNumber, chart, year,
                                    chartsItems.size(), scrapeMillies, dbMillies);
                        }
                    }
                }
            } catch (FailedHostException e) {
                LOGGER.error("Exception during scraping", e);
                // We do not want holes in our charts, because at present we have no means to
                // detect and fix them, therefore move on with next chart.
            }
        }
    }

    private short getFirstYearToScrape(MusicChart chart) {
        var highestYearStoredCharts = chartsItemDao.getHighestYearStored(chart);
        return highestYearStoredCharts >= 0 ? highestYearStoredCharts : chart.yearStarted();
    }

    private byte firstWeek(MusicChart chart, short year) {
        var highestWeekStoredForYear = chartsItemDao.getHighestWeekStored(chart, year);
        if (highestWeekStoredForYear >= 0) {
            // We don't want to re-scrape and store the latest week already scraped
            return (byte) (highestWeekStoredForYear + 1);
        }
        return chart.yearStarted() == year ? chart.weekStarted() : 1;
    }

    private byte lastWeek(short currentYear, short yearProcessing) {
        return currentYear == yearProcessing ? (byte) LocalDate.now().get(ChronoField.ALIGNED_WEEK_OF_YEAR) : 53;
    }

    private DocumentFetcher newInternetDocumentFetcher(MusicChart chart, short year, byte weekNumber) {
        var urlTemplate = scrapeService.getUrlTemplate(chart)
                .orElseThrow(() -> new IllegalStateException(scrapeService + " does not support chart " + chart));
        return new InternetChartsFetcher(String.format(urlTemplate, chart, year, weekNumber));
    }
}

File: src/main/kotlin/info/rsdev/playlists/services/MusicChartsService.kt
/*
 * Copyright 2018 Red Star Development.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package info.rsdev.playlists.services

import java.time.LocalDate
import java.time.temporal.ChronoField
import java.util.ArrayList

import javax.inject.Inject

import info.rsdev.playlists.domain.ChartsItem
import info.rsdev.playlists.domain.Song
import org.slf4j.LoggerFactory

import info.rsdev.playlists.dao.ChartsItemDao
import java.util.function.Consumer

/**
 * This is the default implementation of a [MusicTitleService]
 */
class MusicChartsService

@Inject
constructor(private val scrapeService: ScrapeService, private val chartsItemDao: ChartsItemDao) : MusicTitleService {

    private val LOGGER = LoggerFactory.getLogger(MusicChartsService::class.java)

    override fun init() {
        if (chartsItemDao.setupStoreWhenNeeded()) {
            LOGGER.info("New datastore created succesfully")
        }
        for (chart in MusicChart.values()) {
            val maxYear = chartsItemDao.getHighestYearStored(chart)
            if (maxYear >= 0) {
                val maxWeek = chartsItemDao.getHighestWeekStored(chart, maxYear)
                LOGGER.warn(String.format("Datastore contains data for %s from %d, week %d", chart, maxYear, maxWeek))
            }
        }
        loadData()
    }

    override fun getChartedSongsForYear(year: Short): List<Song> {
        return chartsItemDao.getReleases(year).toList()
    }

    private fun loadData() {
        val currentYear = LocalDate.now().year.toShort()
        for (chart in scrapeService.supportedCharts) {
            for (year in getFirstYearToScrape(chart)..currentYear) {
                for (weekNumber in firstWeek(chart, year.toShort())..lastWeek(currentYear, year.toShort())) {
                    val chartsItems = scrapeService.scrape(newInternetDocumentFetcher(chart, year.toShort(), weekNumber.toByte()))
                    chartsItems.forEach(Consumer<ChartsItem> { chartsItemDao.saveOrUpdate(it) })
                    LOGGER.info("writing week $weekNumber of $chart $year to persistence layer (${chartsItems.size} songs)")
                }
            }
        }
    }

    private fun getFirstYearToScrape(chart: MusicChart): Short {
        val highestYearStoredCharts = chartsItemDao.getHighestYearStored(chart)
        return if (highestYearStoredCharts >= 0) {
            highestYearStoredCharts
        } else chart.yearStarted
    }

    private fun firstWeek(chart: MusicChart, year: Short): Byte {
        val highestWeekStoredForYear = chartsItemDao.getHighestWeekStored(chart, year)
        if (highestWeekStoredForYear >= 0) {
            return highestWeekStoredForYear
        }
        return if (chart.yearStarted == year) {
            chart.weekStarted
        } else 1
    }

    private fun lastWeek(currentYear: Short, yearProcessing: Short): Byte {
        return if (currentYear == yearProcessing) {
            LocalDate.now().get(ChronoField.ALIGNED_WEEK_OF_YEAR).toByte()
        } else 53
    }

    private fun newInternetDocumentFetcher(chart: MusicChart, year: Short, weekNumber: Byte): DocumentFetcher {
        val urlTemplate = scrapeService.getUrlTemplate(chart)?:
            throw IllegalStateException("$scrapeService does not support chart $[chart]")
        return InternetChartsFetcher(String.format(urlTemplate, chart, year, weekNumber))
    }

}
